core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala

```diff
@@ -49,29 +51,18 @@ private[spark] class JobWaiter[T](
     dagScheduler.cancelJob(jobId)
   }
 
-  override def taskSucceeded(index: Int, result: Any): Unit = synchronized {
-    if (_jobFinished) {
-      throw new UnsupportedOperationException("taskSucceeded() called on a finished JobWaiter")
-    }
-    resultHandler(index, result.asInstanceOf[T])
-    finishedTasks += 1
-    if (finishedTasks == totalTasks) {
-      _jobFinished = true
-      jobResult = JobSucceeded
-      this.notifyAll()
+  override def taskSucceeded(index: Int, result: Any): Unit = {
+    /*
+      The resultHandler call must be synchronized in case resultHandler itself
+      is not thread safe.
+     */
+    synchronized(resultHandler(index, result.asInstanceOf[T]))
```

Comment:
Elsewhere in Spark we don't do comments or synchronized blocks like this. Can you follow this style:
```
// The resultHandler call ...
synchronized {
  resultHandler(...)
}
```